Colcoction/unitys-workshop
# Drawing Effects in Character Cards

## Summary
Bug fixes:
- Fixed a bug where empty line breaks were ignored in body-text-parsing.
- Removed a call to `.trimEnd()` in indent block parsing just in case someone wants to use that for formatting too for some reason.
- Fixed a bug where indent labels would delete the first character of their corresponding text, even if it was part of an actual word.

Refactoring:
- Added bullet points as an indent type. Used the lazy method of large indents on deck cards and small indents on character cards.
- Split`drawCardBodyForDeck` into `parseCardBody()` and `drawBodyText()`.
- Consolidated effect-drawing text so that everything uses `parseCardBody()` and `drawBodyText()`
- Removed all the `colorBlahPhase` globals because now everything should be referencing `PHASE_COLOR_MAP`.
- Refactored recursive calls to `drawCardCanvas()` into a single calculation method, `adjustBoxHeightOffset()` which draws the effects once and uses that to determine the box height offset before actually drawing everything. This initial-draw method also lays the groundwork for automatically centering the contents of environment cards.

## Test Plan

For decks & characters:
1. Add line breaks in the game text input. Verify that those line breaks appear in the rendered card.
2. Add some bullet points. On character cards, verify that the bullets have just a little space on the left-hand side. On deck cards, verify that there's a lot of space on the right-hand side.

For just characters:
1. Write some body text on front & back sides with powers, reactions, phases, bullets, extra newlines, and (don't forget) actual normal-looking text. Verify that the box, heading inside the box, logo, HP, and body text are all drawn at the correct height.
2. Write just 1 line of text as body text. Verify that the box, logo, HP, and body text are all drawn at the minimum height.

## Screenshots

![image](https://github.com/Colcoction/unitys-workshop/assets/8094985/77685135-f805-415d-a606-c5d0cbfcc930)

![image](https://github.com/Colcoction/unitys-workshop/assets/8094985/5fe2be3b-cdd4-4c8a-a410-2aa77d5b3276)

![image](https://github.com/Colcoction/unitys-workshop/assets/8094985/42cc3dc6-c6e2-439c-851a-b57f09e55681)


![image](https://github.com/Colcoction/unitys-workshop/assets/8094985/e31d8a9a-437d-42fc-a58b-1da7be57e65f)

![image](https://github.com/Colcoction/unitys-workshop/assets/8094985/fdfb8ee0-2fbf-4ae5-8d66-694cc8044ab7)

## Reviewers
@Colcoction
@sjzhu
